### PR TITLE
Initial MemoryManager and DiskManager APIs  for query execution + External Sort implementation

### LINKED ISSUE #587: [Epic] Optionally Limit memory used by DataFusion plan
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The basic challenge is that as of today, DataFusion can use an unbounded amount of memory for running a plan and it is neither possible to calculate the memory before hand nor limit the use. 

If DataFusion processes individual partitions that are larger than the available memory system memory, right now it will keep allocating memory from the system until it is killed by the OS or container system. 

Also, when running multiple datafusion plans in the same process, each will consume memory without limit where it may be desirable to reserve / cap memory usage by any individual plan to ensure the plans don't together exceed the system memory budget

Thus, it would be nice if we could give DataFusion's plans a memory budget  which they then stayed under

**Describe the solution you'd like**
1. Add an option to ExecutionConfig that has a “total plan memory budget”
2. Add logic to each node that requires a memory buffer to ensure it stays under the limit.

The operators that can use large amounts of memory today are:
1. Sort
2. Join
3. GroupByHash

There are many potential ways to ensure the limit is respected:
1. (Simplest) error if the budget is exceeded
2. (more complex): employ algorithms that can use secondary storage (e.g. temp files) like sort that spills multiple round of partial sorted results and give a final merge phase for the partition global ordering

**Describe alternatives you've considered**
There are some interesting tradeoffs between “up front allocation” dividing memory up across all operators that would need it and a more dynamic approach.

This is likely something that will require some major efforts over many different issues -- I suggest we use this issue to implement a simple "error if over limit" strategy and then work on more sophisticated strategies subsequently

# Progress tracking
Added Jan 2022: 

Remaining Work
- [x] https://github.com/apache/arrow-datafusion/pull/1526
- [ ] https://github.com/apache/arrow-datafusion/issues/3941
- [x] #1568
- [ ] #1570
- [ ] #1599
- [x] #1569
- [x] https://github.com/apache/arrow-datafusion/issues/1637
- [x] https://github.com/apache/arrow-datafusion/issues/4816

### PR DESCRIPTION
**Pull request by yjshen**
# Which issue does this PR close?

Closes #587 .

 # Rationale for this change

When DataFusion processes a single partition, it will keep allocating memory until the OS or the container system kills it. To make it worse, concurrently executing partitions or even simultaneously running plans will compete for available memory until all memory is exhausted. It is more challenging to meet the memory requirements for all operators of each partition when it is running. None of the partitions or plans would run to finish. 

Therefore, the ability to control the total memory usage of the process as a whole, and at the same time, allocate the available memory to each execution partition is extremely important. Under this guarantee: when the memory is sufficient, the operator can acquire as much of the memory to do the computation; when the memory is tight, the operator can be downgraded to use the disk to store some intermediate results (spilling to disk) and use a limited memory for execution.

# What changes are included in this PR?

The proposed memory management architecture is the following:

1. User designates max execution memory by setting `RuntimeConfig.max_memory` and `RuntimeConfig.memory_fraction` (float64 between 0..1). The actual max memory DataFusion could use `pool_size =  max_memory * memory_fraction`.
2. The entities that take up memory during its execution are called `Memory Consumers`. Operators or others are encouraged to register themselves to the memory manager and report its usage through `mem_used()`.
3. There are two kinds of consumers:
   - `Controlling` consumers that would acquire memory during its execution and release memory through `spill` if no more memory is available.
   - `Tracking` consumers that exist for reporting purposes to provide a more accurate memory usage estimation for memory consumers.
4. Controlling and tracking consumers share the pool. Each controlling consumer could acquire a maximum of
   `(pool_size - all_tracking_used) / active_num_controlling_consumers`.

```
            Memory Space for the DataFusion Lib / Process of `pool_size`
   ┌──────────────────────────────────────────────z─────────────────────────────┐
   │                                              z                             │
   │                                              z                             │
   │               Controlling                    z          Tracking           │
   │            Memory Consumers                  z       Memory Consumers      │
   │                                              z                             │
   │                                              z                             │
   └──────────────────────────────────────────────z─────────────────────────────┘

```

# Are there any user-facing changes?

Users could limit the max memory used for DataFusion through `RuntimeConfig::max_memory` and `RuntimeConfig::memory_fraction`

# Note

In addition to the proposed memory manager as well as the runtime that plumbing the execute API, an `ExternalSortExec` is implemented to illustrate the API usage.

### CODE REVIEW

**Comment by alamb** [comment]
Thanks @yjshen  -- I'll try and give this a good look over the weekend

File: datafusion/src/execution/memory_manager.rs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            let sum = trackers.values().fold(0usize, |acc, y| acc + y.mem_used());
+            drop(trackers);
+            self.trackers_total_usage.store(sum, Ordering::SeqCst);
+        }
+    }
+
+    /// Initialize
+    pub(crate) fn initialize(self: &Arc<Self>) {
+        let manager = self.clone();
+        let handle = task::spawn(async move {
```
Comment (houqp) [inline datafusion/src/execution/memory_manager.rs:190]: Am i correct that this background refresh process is needed because tracking consumer memory updates are managed internally instead of through MemoryManager?

File: datafusion/src/execution/memory_manager.rs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+static mut CONSUMER_ID: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    unsafe { CONSUMER_ID.fetch_add(1, Ordering::SeqCst) }
+}
+
+/// Type of the memory consumer
+pub enum ConsumerType {
+    /// consumers that can grow or shrink its memory usage during execution
```
Comment (houqp) [inline datafusion/src/execution/memory_manager.rs:40]: nitpick, the wording might be a little bit misleading. it looks like tracking consumers can also shrink and grow its memory usage as well through heap allocation and deallocation? or do we expect Tracking consumers to only grow used memory at runtime?

File: datafusion/src/execution/memory_manager.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    /// Grow memory attempt from a consumer, return if we could grant that much to it
+    async fn try_grow(
+        self: &Arc<Self>,
+        required: usize,
+        current: usize,
+        consumer_id: &MemoryConsumerId,
+    ) -> bool {
+        let max_per_op = {
+            let total_available =
+                self.pool_size - self.trackers_total_usage.load(Ordering::SeqCst);
```
Comment (houqp) [inline datafusion/src/execution/memory_manager.rs:224]: due to background refresh delay, this is best memory usage counting right? if so, it's worth documenting.

File: datafusion/src/execution/memory_manager.rs (hunk 1 above)
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:190]: Yes, I'm creating a background task that runs periodically to update `tracking consumers'` total memory usage, to avoid `controlling consumers` to ask for available memory frequently.

File: datafusion/src/execution/memory_manager.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        });
+        let _ = self.join_handle.lock().unwrap().insert(handle);
+    }
+
+    /// Register a new memory consumer for memory usage tracking
+    pub(crate) fn register_consumer(self: &Arc<Self>, consumer: Arc<dyn MemoryConsumer>) {
+        let id = consumer.id().clone();
+        match consumer.type_() {
+            ConsumerType::Controlling => {
+                let mut consumers = self.consumers.lock().unwrap();
```
[the remark this answers is not in the document]
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:205]: I'm a little short of naming 😂 . What do you think are proper words that describe the two kinds of consumers?
The current `Controlling / controller` and `Tracking / tracker` seems less informative.

File: datafusion/src/execution/memory_manager.rs (hunk 2 above)
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:40]: I use controlling consumers to describe consumers like external-sorter, hash-aggregator/joiner, or such pipeline breaker operations' consumers that would use a huge amount of memory, and we could control their memory usage pattern by forcing them to spill sometime when no more memory is available. 

Tracking consumers now describe a category of entities that may use some memory during execution, but we do not control them, just check their usage updates periodically. They are free to grow or shrink memory usage, the only demand for the implementors is to be accurate as possible for the `mem_used()` report.

Tracking consumers would occupy tiny memory space, such as a constant internal buffer for each batch calculation, or much larger memory if they are coming from the previous `controlling consumers`.  For example, the last part of in-memory batches for the `external sort` or the already built, partial memory-residential hashtables for hash join probing are tracking consumers we would like to only remember its memory usage, but not trigger its spill to reduce system complexity.

Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:40]: Again, do you think the two`controlling` and `tracking` categories and their usage pattern make sense to you? Any suggestions on the naming?

**Comment by liukun4515** [comment]
@yjshen  thanks, this is milestone pr for memory controller in datafusion.

File: datafusion/src/execution/memory_manager.rs (hunk 2 above)
Comment (houqp) [reply datafusion/src/execution/memory_manager.rs:40]: Ha, ok, this explanation makes a lot more sense to me now. When I initially read the comment, I was thinking `consumers themselves can grow or shrink`. So perhaps we can just reword it to `consumers that can be controlled by memory manager to grow and shrink`.

File: datafusion/src/execution/memory_manager.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (houqp) [reply datafusion/src/execution/memory_manager.rs:205]: Controlling and Tracking seems fine to me. Perhaps we can wait for feed backs from other reviewers.

File: datafusion/src/execution/memory_manager.rs (hunk 1 above)
Comment (houqp) [reply datafusion/src/execution/memory_manager.rs:190]: I haven't put much thought into this yet, but I am curious what are your thoughts on having tracking consumers also report memory usage update directly to the memory manager? Basically similar to what we have with the controlling consumers, but without the capability to force them to spill.

Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:190]: The main reason is to reduce interaction with the memory manager during one's execution, to reduce complexity as well as eliminate synchronization needs.

- For tracking consumers that were converted from controlling consumers. for example, the hashtable size / partial sort in-mem size is known when tracking consumer is created or transformed to, then no more need for them to acquire memory or interact with memory manager. 
- For other tracking consumers with internal computational buffers. One can report its usage by simply updating its internal state `mem_used`, no extra function calls or interaction with the memory manager during execution.

File: datafusion/src/execution/memory_manager.rs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            loop {
+                interval_timer.tick().await;
+                manager.update_tracker_total();
+            }
+        });
+        let _ = self.join_handle.lock().unwrap().insert(handle);
+    }
+
+    /// Register a new memory consumer for memory usage tracking
+    pub(crate) fn register_consumer(self: &Arc<Self>, consumer: Arc<dyn MemoryConsumer>) {
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:201]: I didn't see any code that registered any Tracking consumers yet. 

In terms of plumbing, what do you think about:
1. making all `ExecutionPlans` `MemoryConsumers` and providing default implementations (that reported 0 usage)
2. Registering all ExecutionPlans somehow as MemoryConsumers as part of physical plan creation?

That way all implementations of ExecutionPlan could report their usage without having to explicitly register themselves with the memory manager. Also the manager could report on how many operators were not providing any statistics, etc

File: datafusion/src/execution/memory_manager.rs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+    /// Drop a memory consumer from memory usage tracking
+    pub(crate) fn drop_consumer(self: &Arc<Self>, id: &MemoryConsumerId) {
+        // find in consumers first
+        {
+            let mut consumers = self.consumers.lock().unwrap();
+            if consumers.contains_key(id) {
+                consumers.remove(id);
+                return;
+            }
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:247]:
```suggestion
            if let Some(_) = consumers.remove(id) {
                return;
            }
```

File: datafusion/src/execution/memory_manager.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            if trackers.contains_key(id) {
+                let removed = trackers.remove(id);
+                match removed {
+                    None => {}
+                    Some(tracker) => {
+                        let usage = tracker.mem_used();
+                        self.trackers_total_usage.fetch_sub(usage, Ordering::SeqCst);
+                    }
+                }
+            }
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:260]:
```suggestion
            if let Some(tracker) = trackers.remove(id) {
                        let usage = tracker.mem_used();
                        self.trackers_total_usage.fetch_sub(usage, Ordering::SeqCst);
            }
```

File: datafusion/src/execution/memory_manager.rs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    /// Shutdown
+    pub(crate) fn shutdown(&mut self) {
+        let maybe_handle = self.join_handle.lock().unwrap().take();
+        match maybe_handle {
+            None => {}
+            Some(handle) => handle.abort(),
+        }
+    }
+
+    #[allow(dead_code)]
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:273]: Recommend making this `impl Display for MemoryManager` instead

File: datafusion/src/execution/memory_manager.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+   │               Controlling                    z          Tracking           │
+   │            Memory Consumers                  z       Memory Consumers      │
+   │                                              z                             │
+   │                                              z                             │
+   └──────────────────────────────────────────────z─────────────────────────────┘
+*/
+
+/// Manage memory usage during physical plan execution
+pub struct MemoryManager {
+    consumers: Arc<Mutex<HashMap<MemoryConsumerId, Arc<dyn MemoryConsumer>>>>,
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:155]: As mentioned above, I think there is a refcount cycle here

Specifically, `ExternalSort` has a `RuntimeEnv` which has an `Arc<MemoryManager>` but the MemoryManager has a `Arc` back to the `ExternalSort`.

Thus I suggest changing this and the trackers below to be

```suggestion
    consumers: Arc<Mutex<HashMap<MemoryConsumerId, Weak<dyn MemoryConsumer>>>>,
```

to break cycle

**Review by alamb — COMMENTED** [review]
Thank you so much @yjshen 

I think the memory manager interface in this PR (`MemoryManager` / `MemoryConsumer`)  is a nice good foundation going forward. 

Prior to merging this PR I would like to see:
1. The ref count cycle between the memory manager and execution plans, I think this PR could me merged into DataFusion as is and we could iterate from there
2. Some tests for `MemoryManager` and `ExternalSorter`  (as suggested in the PR comments here)

I also think it is worth removing / reconsidering the background loop for tracked memory consumers as well, though since there isn't used yet I don't think it is critical to remove prior to merging this PR

But again, really nice and thank you for this contribution

File: datafusion/src/execution/memory_manager.rs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+        self.id().partition_id
+    }
+
+    /// Type of the consumer
+    fn type_(&self) -> &ConsumerType;
+
+    /// Grow memory by `required` to buffer more data in memory,
+    /// this may trigger spill before grow when the memory threshold is
+    /// reached for this consumer.
+    async fn try_grow(&self, required: usize) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs:94]: Maybe it's `try_grow` in both `MemoryManager` and `MemoryConsumer` misleading. 

In MemoryConsumer, The above three `try_grow()` `spill()` `memory_manager()`  act as follows:
`try_grow()` asks for more memory from `memory_manger()` , if not granted, call `spill()` on itself.  

In MemoryManger, I renamed the function `can_grow()`:
`can_grow` first gathers all tracker consumers' total memory, then the active number of memory requesters, at last, it tells if the requesters could be fulfilled.

File: datafusion/src/execution/memory_manager.rs (hunk 1 above)
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs]: There are no more maintained `tracker_total` and no more background maintaining tasks. Memory manager decides total tracker memory each time its `can_grow` is called now.

**Comment by yjshen** [comment]
@houqp @alamb  Thanks for your detailed and insightful review!

### Resolved:

- The maintained total trackers' memory and the background threads that update it are removed. Instead, total tracker memory is collected each time the memory manager runs its `can_grow` method.
- Renamed controlling consumer to requesting consumer, requester in short.
- Use Weak<dyn MemoryConsumer> in MemoryManager now.
- Use TempDir instead of manually retries of creating scratch dirs. Use rand instead of `uuid` crate.

### To discuss:
> I didn't see any code that registered any Tracking consumers yet.

There is one in `SortMergingStream` while merging multiple partial order results from spill files and the last piece of batches that are still in memory. The last piece is created as in-memory batches backed `StreamWrapper`, and just reporting its usage as in-memory batches total size.
> 
> In terms of plumbing, what do you think about:
> 
> 1. making all `ExecutionPlans` `MemoryConsumers` and providing default implementations (that reported 0 usage)
> 2. Registering all ExecutionPlans somehow as MemoryConsumers as part of physical plan creation?
> 
> That way all implementations of ExecutionPlan could report their usage without having to explicitly register themselves with the memory manager. Also the manager could report on how many operators were not providing any statistics, etc

I think there is a gap between ExecPlan and MemoryConsumer.  Since an `execute` method would be called multiple times with different `partition`, it's always the `SendableRecordBatchStream` such as `SortPreservingMergeStream`, `CrossJoinStream` that takes up memory. Should I make it like:

```rust

/// Trait for types that stream [arrow::record_batch::RecordBatch]
pub trait RecordBatchStream: Stream<Item = ArrowResult<RecordBatch>> + MemoryConsumer {
    /// Returns the schema of this `RecordBatchStream`.
    ///
    /// Implementation of this trait should guarantee that all `RecordBatch`'s returned by this
    /// stream should have the same schema as returned from this method.
    fn schema(&self) -> SchemaRef;
}

/// Trait for a stream of record batches.
pub type SendableRecordBatchStream = Pin<Arc<dyn RecordBatchStream + Send + Sync>>;

```
Should I make SendableRecordBatchStream pin arc instead of pin box and register each stream arc to runtime at each `execute()` last line? Also register consumers through:

```
    pub fn register_consumer(&self, memory_consumer: &Arc<dyn MemoryConsumer>) {
``
may sometimes be awkward:

```
        runtime.register_consumer(&(streams.clone() as Arc<dyn MemoryConsumer>));
``` 

Any thoughts?

**Comment by alamb** [comment]
> I think there is a gap between ExecPlan and MemoryConsumer. Since an execute method would be called multiple times with different partition, it's always the SendableRecordBatchStream such as SortPreservingMergeStream, CrossJoinStream that takes up memory. Should I make it like:

This is a good point (that the memory management is done on a per-partition basis rather than a per `ExecutionPlan` basis. I need to think 🤔  about it some more. 

I would recommend we don't change `SendableRecordBatchStream` which is complicated enough as is. 

I will make time today to review this PR again thoroughly -- thank you @yjshen  I think we are close

**Comment by yjshen** [comment]
@tustvold Thanks for bringing it up. I find the stream a single place to have all runtime entities be auto-registered to the memory manager at once. Maybe a wrapper over the stream could achieve the goal?

**Comment by tustvold** [comment]
> Maybe a wrapper over the stream could achieve the goal

My instinct would be to suggest having the shared ref internal to the stream implementation, instead of a wrapper. Otherwise I suspect you will run into borrow checker, and async pain. This would also avoid needing to make breaking changes to SendableRecordBatchStream?

Another thing to potentially think about is that many of the operators aren't actually streams, rather they spawn a tokio task and then return an mpsc queue. There will need to be some accounting of both data buffered in the queue, and data in the operators "task". My gut feeling is this is going to require adding some sort of RAII tracking field to `RecordBatch` or possibly `Buffer` but I'm not really sure...

File: datafusion/src/execution/memory_manager.rs [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+use std::fmt;
+use std::fmt::{Debug, Display, Formatter};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Condvar, Mutex, Weak};
+
+static mut CONSUMER_ID: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    unsafe { CONSUMER_ID.fetch_add(1, Ordering::SeqCst) }
+}
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:33]: I don't think we need mut / unsafe here. This works for me locally; 
```suggestion
static CONSUMER_ID: AtomicUsize = AtomicUsize::new(0);

fn next_id() -> usize {
    CONSUMER_ID.fetch_add(1, Ordering::SeqCst)
}
```

File: datafusion/src/physical_plan/sorts/in_mem_sort.rs [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+use crate::physical_plan::metrics::BaselineMetrics;
+use crate::physical_plan::sorts::{RowIndex, SortKeyCursor};
+use crate::physical_plan::{
+    expressions::PhysicalSortExpr, PhysicalExpr, RecordBatchStream,
+};
+
+/// Merge buffered, self-sorted record batches to get an order.
+///
+/// Internally, it uses MinHeap to reduce extra memory consumption
+/// by not concatenating all batches into one and sorting it as done by `SortExec`.
```
Comment (alamb) [inline datafusion/src/physical_plan/sorts/in_mem_sort.rs:42]: I need to study the connection between `SortExec`, `SortPreservingMergeStream` and `InMemSortStream` some more to fully get understand this. I can't help by think that `InMemSortStream` is doing the same thing as `SortPreservingMergeStream` -- and I wonder if we can reuse that same code

**Review by alamb — COMMENTED** [review]
I reviewed the memory manager changes -- I think it is good enough to start with and we can iterate from there.

I didn't get a chance to fully review the changes to `sort_preserving_merge` -- will keep at it tomorrow.

cc @tustvold

File: datafusion/src/physical_plan/sorts/in_mem_sort.rs (hunk 12 above)
Comment (yjshen) [reply datafusion/src/physical_plan/sorts/in_mem_sort.rs:42]: The main difference between `InMemSortStream` and `SortPreservingMergeStream` lies in assuming different numbers of "entity" (batches for IMSS and streams for SPMS) merged. 

Since `InMemSort` means to merge much more partially ordered "entities", the sorter should reduce the num of comparison for each item pop-up. Hence a MinHeap was introduced.

On the other hand, `InMemSort` is more specialized to have each "entity" only one record batch, therefore simplified logic compared to consider stream continuation in `SortPreservingMergeStream`.

Currently, the common parts `SortKeyCursor` and `RowIndex` for both sorts are extracted to `sorts/mod.rs` reduce duplication.

File: datafusion/src/execution/memory_manager.rs [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+        if !can_grow_directly {
+            info!(
+                "Failed to grow memory of {} directly from consumer {}, spilling first ...",
+                human_readable_size(required),
+                self.id()
+            );
+            let freed = self.spill().await?;
+            self.memory_manager().record_free(freed);
+        }
+        self.memory_manager().record_acquire(required);
```
Comment (houqp) [inline datafusion/src/execution/memory_manager.rs:114]: `request_total` could change after we called `can_grow_directly` and before we call `record_acquire` here. If we are doing best effort accounting, then this is not the end of the world, but just wanted to point this out in case it's not expected.

**Review by houqp — APPROVED** [review]
I also don't have a good solution to the `SendableRecordBatchStream` problem off the top of my head, will need to think more about it. Other than that, I think the change looks good as a first iteration :+1:

File: datafusion/src/execution/memory_manager.rs (hunk 13 above)
Comment (yjshen) [reply datafusion/src/execution/memory_manager.rs]: Fixed, mind taking another look?

**Review by alamb — APPROVED** [review]
Thank you again @yjshen for this great contribution. 

I have some concerns about parts of this PR (listed below), but I still think we should merge this PR as is and handle the concerns as follow ons.

Thus my plan is to send a note to the dev mailing list and slack channel asking for comments (on the API specifically) and if we don't hear any major concerns I suggest we merge this PR tomorrow.

My rationale for the merge with concerns doing is:
1. This PR gets the necessary foundations in for limiting resources at runtime: specifically the `RuntimeEnv`, and `MemoryConsumer`, `MemoryManager`, and `DiskManager` APIs.
2. It is backwards compatible (e.g. external sort is not connected to anything, so there should be no performance regressions)

I think it would be good to file a follow on PR marking the MemoryManager, DiskManager, and MemoryConsumer APIs as experimental and I will prepare such a PR.

My Concerns (aka major follow on work):
1. External sorting is not connected to anything -- aka it is code that isn't used (yet)
2. `InMemSortStream` and `SortPreservingMergeStream` are doing very much the same thing -- consolidating the code I think will be important as we move to optimize them
3. As most of the rest of the system isn't connected to the memory system, the APIs may not be fully adequate (but we can iterate on that)


This PR also unlocks some cool follow on projects (like supporting external group by / group by hash / spill to disk) 🚗

File: datafusion/src/execution/runtime_env.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    /// Runtime configuration
+    pub config: RuntimeConfig,
+    /// Runtime memory management
+    pub memory_manager: Arc<MemoryManager>,
+    /// Manage temporary files during query execution
+    pub disk_manager: Arc<DiskManager>,
+}
+
+impl Debug for RuntimeEnv {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
```
Comment (xudong963) [inline datafusion/src/execution/runtime_env.rs:39]: Why use anonymous lifetime? Maybe it's unnecessary.

File: datafusion/src/execution/runtime_env.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    }
+}
+
+impl Default for RuntimeConfig {
+    fn default() -> Self {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let path = tmp_dir.path().to_str().unwrap().to_string();
+        std::mem::forget(tmp_dir);
+
+        Self {
```
Comment (xudong963) [inline datafusion/src/execution/runtime_env.rs:141]: It is better to define constants for default values of `batch_size` and `memory_fraction` than bare numbers

**Review by xudong963 — COMMENTED** [review]
Thanks, @yjshen, great work! I went through it briefly.  BTW, some of the details are well handled, such as `pool size`.

File: datafusion/src/execution/memory_manager.rs (hunk 13 above)
Comment (houqp) [reply datafusion/src/execution/memory_manager.rs]: fix looks good to me :+1:

File: datafusion/src/execution/runtime_env.rs (hunk 14 above)
Comment (yjshen) [reply datafusion/src/execution/runtime_env.rs:39]: Ah, my IDE generates it. I think it aligns with the method signature and the doc. Is it recommended to remove?

Comment (xudong963) [reply datafusion/src/execution/runtime_env.rs:39]: The examples in https://doc.rust-lang.org/core/fmt/struct.Formatter.html don't add anonymous lifetime. Add and unadd both can be compiled. I'm just curious why you added it here. I don't have a strong preference between the two.

File: datafusion/src/execution/runtime_env.rs (hunk 15 above)
Comment (yjshen) [reply datafusion/src/execution/runtime_env.rs:141]: I think this is already the default of config and meant to be overwritten if needed. And a single place for these defaults? 

BTW, I think it is worth reconsidering and restructuring the multiple configs and their usages. `ExecutionConfig`, `PhysicalPlanConfig` and `RuntimeConfig`. 

At least we should not pass `target_batch_size` during query planning since we already have runtimeEnv plumbing through the execute() API now, will create follow-up PR once we've merged this one.

**Comment by liukun4515** [comment]
> FWIW I also plan to run the TPCH benchmarks on this PR and will post the results (I don't expect any changes)

You will post the result in this review?

**Comment by liukun4515** [comment]
> FWIW I also plan to run the TPCH benchmarks on this PR and will post the results (I don't expect any changes)

Will you post the result in this review?

File: datafusion/src/execution/runtime_env.rs (hunk 15 above)
Comment (xudong963) [reply datafusion/src/execution/runtime_env.rs:141]: Make sense to me. We can deal with it uniformly in the following PR

File: datafusion/src/execution/mod.rs [hunk 16]
```diff
@@ -19,4 +19,7 @@
 
 pub mod context;
 pub mod dataframe_impl;
+pub mod disk_manager;
```
Comment (liukun4515) [inline datafusion/src/execution/mod.rs:22]: Do we need to public this api for the lib?

Comment (yjshen) [reply datafusion/src/execution/mod.rs:22]: I may need to expose df generated dirs to outer modules, for instance, write shuffle data in df in stage-1 and read shuffle data in reducers run in JVM. I think it's no harm to expose the module out?

Comment (liukun4515) [reply datafusion/src/execution/mod.rs:22]: If we public more API, we should take more effort to maintain the api back compatibility feature.

**Review by xudong963 — APPROVED** [review]

File: datafusion/src/execution/mod.rs (hunk 16 above)
Comment (yjshen) [reply datafusion/src/execution/mod.rs]: Thanks, I've changed these two into `pub(crate)`

File: datafusion/src/execution/runtime_env.rs (hunk 14 above)
Comment (yjshen) [reply datafusion/src/execution/runtime_env.rs]: Thanks @xudong963 I've removed the anonymous lifetime in the lastest commit

**Comment by alamb** [comment]
> Will you post the result in this review?

Yes, I will do so.

**Comment by alamb** [comment]
Ok, I am going to fire up my benchmark machine, get some numbers, and assuming they look good merge this PR

**Comment by alamb** [comment]
I think while not perfect this PR is a step in the right direction  towards being able to handle queries that need to spill. 🚀  thank you @yjshen 

Shall I file follow on tickets for the next step?

I am particularly interested in ensuring we consolidate the Sort code (so there is only a single sort operator that does in memory sorting if it has enough memory budget but then spills to disk if needed). I would enjoy helping make this happen (perhaps by writing some tests?)

**Comment by liukun4515** [comment]
> 
I talked with @yjshen yesterday, maybe the external sort has been implemented as a draft in his branch. @alamb 
You can fill an issue to track the follow-up tasks, it's clear for all contributors.

**Comment by yjshen** [comment]
Thank you all again for helping me with the initial document proposal as well as insightful reviews in this PR ❤️

> Shall I file follow on tickets for the next step?

Yes, that would be great! please open issues that you have in mind. I have a bunch of ideas as follow-ups as well. I think we already have the foundation for many exciting features to come. How do you like to open an umbrella issue as well as sub-task issues? I could file sub-tasks issues under it as well.

> I am particularly interested in ensuring we consolidate the Sort code.  (so there is only a single sort operator that does in memory sorting if it has enough memory budget but then spills to disk if needed). I would enjoy helping make this happen (perhaps by writing some tests?)

Thanks, I can do the initial consolidation, please join at any time or just take it over. depends on your time schedule.
